### PR TITLE
Dataset and Dataloader classes to work with Hugging Face transformers…

### PR DESCRIPTION
**Pull request by albertogaspar**
Added dataloaders classes to deal with inputs sources that are not just plain `jax.Array`. In particular, in this PR I added the capabilities of dealing with text sequences using the `transformers` library by HuggingFace.
This PR focuses on adding support for the Sequence Classification use case.

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Fortuna can only deal with inputs batches like `(X,y)` where `X` and `y` (inputs and targets, respectively) are `jax.Array`

## What is the new behavior?

Added new Dataloader to deal with inputs batches like `(X,y)` where `X` is a `Dict[str, jax.Array]` and `y` is a `jax.Array`.

### CODE REVIEW

File: fortuna/data/dataset/huggingface_datasets.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        self,
+        rng: PRNGKeyArray,
+        tokenizer: PreTrainedTokenizer,
+        max_length: int,
+        padding: Union[bool, str, PaddingStrategy] = True,
+        per_device_train_batch_size: int = 8,
+        per_device_eval_batch_size: int = 8
+    ):
+        """
+        A Dataset class to work with Hugging-Face like datasets.
```
Comment (gianlucadetommaso) [inline fortuna/data/dataset/huggingface_datasets.py:31]: Perhaps remove "like"? Or add "-" b/w HF and like.

File: fortuna/data/dataset/huggingface_datasets.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+
+class HuggingFaceClassificationDatasetABC(abc.ABC):
+    def __init__(
+        self,
+        rng: PRNGKeyArray,
+        tokenizer: PreTrainedTokenizer,
+        max_length: int,
+        padding: Union[bool, str, PaddingStrategy] = True,
+        per_device_train_batch_size: int = 8,
```
Comment (gianlucadetommaso) [inline fortuna/data/dataset/huggingface_datasets.py:27]: Is 8 a standard default?

File: fortuna/data/dataset/huggingface_datasets.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        padding: Union[bool, str, PaddingStrategy]
+            See `Padding and Truncation <https://huggingface.co/docs/transformers/pad_truncation>`_
+            for more information (`truncation` is always True).
+        per_device_train_batch_size: int
+            Batch size for each device to be used for training
+        per_device_eval_batch_size: int
+            Batch size for each device to be used for evaluation
+        """
+        rngs = jax.random.split(rng, 3)
+        self.rngs = {'train': rngs[0], 'validation': rngs[1], 'test': rngs[2]}
```
Comment (gianlucadetommaso) [inline fortuna/data/dataset/huggingface_datasets.py:51]: In our framework, we may also want to hold out an extra data set for calibration. Would it make sense to make it appear in the split?

File: pyproject.toml [hunk 4]
```diff
@@ -15,6 +15,8 @@ optax = "^0.1.3"
 matplotlib = "^3.6.2"
 tqdm = "^4.64.1"
 numpy = "^1.23.4"
+transformers = "^4.28.1"
```
Comment (gianlucadetommaso) [inline pyproject.toml:18]: Question, open to discussion. Do we want to depend on transformers by default? Or should we rather have it as an optional dependency, to enable if the user wants to use Hugging Face models and datasets?

File: fortuna/data/dataset/huggingface_datasets.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            batch = dataset[batch_idxs]
+            batch = self._collate(batch)
+            batch_inputs = {k: jnp.array(v) for k, v in batch.items()}
+            if 'labels' in batch_inputs:
+                batch_labels = batch_inputs.pop('labels')
+                yield (batch_inputs, batch_labels)
+            else:
+                yield batch_inputs
+
+    def get_train_dataloader(self, dataset: Dataset) -> HuggingFaceDataLoader:
```
Comment (gianlucadetommaso) [inline fortuna/data/dataset/huggingface_datasets.py:127]: Related to the comment above, should we consider also offering a `get_val_data_loader` and `get_calib_data_loader`?

Or, more generally, should we consider offering a generic `get_data_loader`, where the user can decide whether to shuffle or not (for example)? I can imagine for output calibration perhaps one might want a data loader that does not shuffle.

**Review by gianlucadetommaso — COMMENTED** [review]
Neat implementation :) Just a couple of comments and questions.

File: pyproject.toml (hunk 4 above)
Comment (albertogaspar) [reply pyproject.toml:18]: Having it as an optional-dependency is probably a better option

File: fortuna/data/dataset/huggingface_datasets.py (hunk 2 above)
Comment (albertogaspar) [reply fortuna/data/dataset/huggingface_datasets.py:27]: I think it's a reasonable default

File: fortuna/data/dataset/huggingface_datasets.py (hunk 3 above)
Comment (albertogaspar) [reply fortuna/data/dataset/huggingface_datasets.py:51]: I'll refactor this

File: fortuna/data/loader/huggingface_loaders.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+from fortuna.typing import Array
+
+
+class HuggingFaceDataLoader(BaseDataLoaderABC):
+
+    def __init__(
+            self,
+            iterable: Union[Iterable[Dict[str, Array]], Iterable[Tuple[Dict[str, Array], Array]]],
+            num_unique_labels: int = None,
+            num_input_samples: Optional[int] = None
```
Comment (gianlucadetommaso) [inline fortuna/data/loader/huggingface_loaders.py:15]: Can you add these two arguments to the docstring?

Also, what do you mean. by samples? If this is just the number of inputs, I'd just call it `num_inputs`. I'd prefer avoiding the confusion of what a "sample" is.

File: fortuna/data/dataset/huggingface_datasets.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+    @property
+    @abc.abstractmethod
+    def data_collator(self):
+        """
+        Create a data collator object that will process batches of data and apply transformation on them if needed.
+        """
+        pass
+
+    def get_dataloader(self, dataset: Dataset,  rng: PRNGKeyArray, train: bool = False, shuffle: bool = False, drop_last: bool = False, verbose: bool = False) -> HuggingFaceDataLoader:
```
Comment (gianlucadetommaso) [inline fortuna/data/dataset/huggingface_datasets.py:89]: Can we call it `get_data_loader`, similarly to the private method?

File: fortuna/data/dataset/huggingface_datasets.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        if drop_last:
+            steps_per_epoch = dataset_size // batch_size
+            dataset_idxs = dataset_idxs[: steps_per_epoch * batch_size]  # Skip incomplete batch.
+        yield from [dataset_idxs[start:start+batch_size] for start in range(0, len(dataset_idxs), batch_size)]
+
+    def _get_data_loader(
+            self, dataset: Dataset,  batch_size: int, shuffle: bool = False, drop_last: bool = False,
+            rng: Optional[PRNGKeyArray] = None, verbose: bool = False,
+    ) -> Union[Iterable[Dict[str, Array]], Iterable[Tuple[Dict[str, Array],Array]]]:
+        rng = jax.random.split(rng, 1)[0]
```
Comment (gianlucadetommaso) [inline fortuna/data/dataset/huggingface_datasets.py:148]: Why doing a split here?

File: fortuna/data/dataset/huggingface_datasets.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+
+class HuggingFaceClassificationDatasetABC(abc.ABC):
+    def __init__(
+        self,
+        tokenizer: PreTrainedTokenizer,
+        max_length: int,
+        padding: Union[bool, str, PaddingStrategy] = True,
+        per_device_train_batch_size: int = 8,
+        per_device_eval_batch_size: int = 8,
```
Comment (gianlucadetommaso) [inline fortuna/data/dataset/huggingface_datasets.py:27]: Is this `train` and `eval` distinction still relevant now that you don't offer any longer explicit training and test loaders?

Comment (gianlucadetommaso) [reply fortuna/data/dataset/huggingface_datasets.py:27]: Would it be better to offer just one, and the user calls the object for each data loader (train, val, calib, test)?

Comment (albertogaspar) [reply fortuna/data/dataset/huggingface_datasets.py:27]: Yes the distinction is relevant because in `eval` you can use larger bach size given that there is no backward pass.
